Flaky-test: SimpleProducerConsumerTest.testConcurrentConsumerReceiveWhileReconnect

org.apache.pulsar.client.api.SimpleProducerConsumerTest is flaky. The testConcurrentConsumerReceiveWhileReconnect test method fails sporadically.

```
java.lang.AssertionError: expected [10] but found [0]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.client.api.SimpleProducerConsumerTest.testConcurrentConsumerReceiveWhileReconnect(SimpleProducerConsumerTest.java:831)
```

Usage tip: To enable automatic navigation to failure message, open the links with CTRL/CMD-click.
[example failure 2021-02-02T13:43:07.5573972Z](https://github.com/apache/pulsar/runs/1813006661?check_suite_focus=true#step:9:242)
[example failure 2021-02-02T13:43:07.5535772Z](https://github.com/apache/pulsar/runs/1813006661?check_suite_focus=true#step:9:220)
[example failure 2021-02-02T08:31:11.6255612Z](https://github.com/apache/pulsar/runs/1812415717?check_suite_focus=true#step:8:220)
[example failure 2021-02-01T14:30:09.0740501Z](https://github.com/apache/pulsar/runs/1806917870?check_suite_focus=true#step:9:220)
[example failure 2021-02-01T04:00:30.6334452Z](https://github.com/apache/pulsar/runs/1803769725?check_suite_focus=true#step:9:220)
[example failure 2021-01-31T01:38:11.4471620Z](https://github.com/apache/pulsar/runs/1799451053?check_suite_focus=true#step:9:391)
[example failure 2021-01-31T01:38:11.4351107Z](https://github.com/apache/pulsar/runs/1799451053?check_suite_focus=true#step:9:369)
[example failure 2021-01-31T01:37:47.7379447Z](https://github.com/apache/pulsar/runs/1799451900?check_suite_focus=true#step:9:392)
[example failure 2021-01-31T01:33:11.4717666Z](https://github.com/apache/pulsar/runs/1799451900?check_suite_focus=true#step:9:242)
[example failure 2021-01-31T01:33:11.4548599Z](https://github.com/apache/pulsar/runs/1799451900?check_suite_focus=true#step:9:220)
[example failure 2021-01-31T01:33:01.4379453Z](https://github.com/apache/pulsar/runs/1799451053?check_suite_focus=true#step:9:220)
[example failure 2021-01-30T18:05:01.0381990Z](https://github.com/apache/pulsar/runs/1798184515?check_suite_focus=true#step:9:519)
[example failure 2021-01-30T17:59:30.5258320Z](https://github.com/apache/pulsar/runs/1798184515?check_suite_focus=true#step:9:369)
[example failure 2021-01-30T17:54:15.7235852Z](https://github.com/apache/pulsar/runs/1798184515?check_suite_focus=true#step:9:220)
[example failure 2021-01-30T15:25:31.1833527Z](https://github.com/apache/pulsar/runs/1797736135?check_suite_focus=true#step:9:370)
[example failure 2021-01-30T15:20:22.0673609Z](https://github.com/apache/pulsar/runs/1797736135?check_suite_focus=true#step:9:220)
[example failure 2021-01-30T14:44:07.9673268Z](https://github.com/apache/pulsar/runs/1797634104?check_suite_focus=true#step:9:220)
[example failure 2021-01-30T14:33:26.3691573Z](https://github.com/apache/pulsar/runs/1797584071?check_suite_focus=true#step:9:391)
[example failure 2021-01-30T14:33:26.3596761Z](https://github.com/apache/pulsar/runs/1797584071?check_suite_focus=true#step:9:369)
[example failure 2021-01-30T14:28:07.6813129Z](https://github.com/apache/pulsar/runs/1797584071?check_suite_focus=true#step:9:220)
[example failure 2021-01-29T16:51:32.6326558Z](https://github.com/apache/pulsar/runs/1791621783?check_suite_focus=true#step:8:519)
[example failure 2021-01-29T16:46:15.9995927Z](https://github.com/apache/pulsar/runs/1791621783?check_suite_focus=true#step:8:370)
[example failure 2021-01-29T16:40:57.6363027Z](https://github.com/apache/pulsar/runs/1791621783?check_suite_focus=true#step:8:220)
[example failure 2021-01-29T09:40:49.3355976Z](https://github.com/apache/pulsar/runs/1790322670?check_suite_focus=true#step:8:242)
[example failure 2021-01-29T09:40:49.3265967Z](https://github.com/apache/pulsar/runs/1790322670?check_suite_focus=true#step:8:220)
[example failure 2021-01-29T01:05:21.7483394Z](https://github.com/apache/pulsar/runs/1788381625?check_suite_focus=true#step:8:220)
[example failure 2021-01-28T16:23:46.4454848Z](https://github.com/apache/pulsar/runs/1784431383?check_suite_focus=true#step:8:369)
[example failure 2021-01-28T16:18:57.9386200Z](https://github.com/apache/pulsar/runs/1784431383?check_suite_focus=true#step:8:220)
[example failure 2021-01-27T05:57:39.9518491Z](https://github.com/apache/pulsar/runs/1774583699?check_suite_focus=true#step:8:370)
[example failure 2021-01-27T05:52:49.9845436Z](https://github.com/apache/pulsar/runs/1774583699?check_suite_focus=true#step:8:220)


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [10] but found [0]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.client.api.SimpleProducerConsumerTest.testConcurrentConsumerReceiveWhileReconnect(SimpleProducerConsumerTest.java:831)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

</pre></code>
</details>

--- a/pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerTest.java
@@ -760,12 +760,9 @@ public class SimpleProducerConsumerTest extends ProducerConsumerBase {
         }
 
         barrier.await();
-        // there will be 10 threads calling receive() from the same consumer and will block
-        Thread.sleep(100);
 
         // we restart the broker to reconnect
         restartBroker();
-        Thread.sleep(2000);
 
         // publish 100 messages so that the consumers blocked on receive() will now get the messages
         ProducerBuilder<byte[]> producerBuilder = pulsarClient.newProducer()
@@ -781,12 +778,14 @@ public class SimpleProducerConsumerTest extends ProducerConsumerBase {
             String message = "my-message-" + i;
             producer.send(message.getBytes());
         }
-        Thread.sleep(500);
 
         ConsumerImpl<byte[]> consumerImpl = (ConsumerImpl<byte[]>) consumer;
-        // The available permits should be 10 and num messages in the queue should be 90
-        Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads);
-        Assert.assertEquals(consumerImpl.numMessagesInQueue(), recvQueueSize - numConsumersThreads);
+
+        Awaitility.await().untilAsserted(() -> {
+            // The available permits should be 10 and num messages in the queue should be 90
+            Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads);
+            Assert.assertEquals(consumerImpl.numMessagesInQueue(), recvQueueSize - numConsumersThreads);
+        });
 
         barrier.reset();
         for (int i = 0; i < numConsumersThreads; i++) {
@@ -797,11 +796,12 @@ public class SimpleProducerConsumerTest extends ProducerConsumerBase {
             });
         }
         barrier.await();
-        Thread.sleep(100);
 
-        // The available permits should be 20 and num messages in the queue should be 80
-        Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads * 2);
-        Assert.assertEquals(consumerImpl.numMessagesInQueue(), recvQueueSize - (numConsumersThreads * 2));
+        Awaitility.await().untilAsserted(() -> {
+            // The available permits should be 20 and num messages in the queue should be 80
+            Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads * 2);
+            Assert.assertEquals(consumerImpl.numMessagesInQueue(), recvQueueSize - (numConsumersThreads * 2));
+        });
 
         // clear the queue
         while (true) {
@@ -811,9 +811,11 @@ public class SimpleProducerConsumerTest extends ProducerConsumerBase {
             }
         }
 
-        // The available permits should be 0 and num messages in the queue should be 0
-        Assert.assertEquals(consumerImpl.getAvailablePermits(), 0);
-        Assert.assertEquals(consumerImpl.numMessagesInQueue(), 0);
+        Awaitility.await().untilAsserted(() -> {
+            // The available permits should be 0 and num messages in the queue should be 0
+            Assert.assertEquals(consumerImpl.getAvailablePermits(), 0);
+            Assert.assertEquals(consumerImpl.numMessagesInQueue(), 0);
+        });
 
         barrier.reset();
         for (int i = 0; i < numConsumersThreads; i++) {
@@ -824,15 +826,14 @@ public class SimpleProducerConsumerTest extends ProducerConsumerBase {
             });
         }
         barrier.await();
-        // we again make 10 threads call receive() and get blocked
-        Thread.sleep(100);
 
         restartBroker();
-        Thread.sleep(2000);
 
-        // The available permits should be 10 and num messages in the queue should be 90
-        Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads);
-        Assert.assertEquals(consumerImpl.numMessagesInQueue(), recvQueueSize - numConsumersThreads);
+        Awaitility.await().untilAsserted(() -> {
+            // The available permits should be 10 and num messages in the queue should be 90
+            Assert.assertEquals(consumerImpl.getAvailablePermits(), numConsumersThreads);
+            Assert.assertEquals(consumerImpl.numMessagesInQueue(), recvQueueSize - numConsumersThreads);
+        });
         consumer.close();
         executor.shutdown();
     }

